xdsl/irdl/constraints.py

```diff
@@ -256,23 +199,16 @@ def verify(self, attr: Attribute, constraint_context: ConstraintContext) -> None
             raise VerifyException(f"attribute {attr} expected to be a TypedAttribute")
         self.type_constraint.verify(attr.get_type(), constraint_context)
 
-    @dataclass(frozen=True)
-    class _Extractor(VarExtractor[Attribute]):
-        inner: VarExtractor[Attribute]
-
-        def extract_var(self, a: Attribute) -> ConstraintVariableType:
-            if not isinstance(a, TypedAttribute):
-                raise PyRDLError(f"Inference expected {a} to be a TypedAttribute")
-            return self.inner.extract_var(a.get_type())
-
-    def get_variable_extractors(self) -> dict[str, VarExtractor[Attribute]]:
-        return merge_extractor_dicts(
-            self.attr_constraint.get_variable_extractors(),
-            {
-                v: self._Extractor(r)
-                for v, r in self.type_constraint.get_variable_extractors().items()
-            },
-        )
+    def variables(self) -> dict[str, ConstraintVarType]:
+        return self.type_constraint.variables() | self.attr_constraint.variables()
+
+    def extract_var(self, attr: Attribute, var: str) -> ConstraintVariableType:
+        if var in self.attr_constraint.variables():
+            return self.attr_constraint.extract_var(attr, var)
+        else:
+            if not isinstance(attr, TypedAttribute):
+                raise PyRDLError(f"Inference expected {attr} to be a TypedAttribute")
+            return self.type_constraint.extract_var(attr.get_type(), var)
```

Comment:
Feels like the pythonic way of doing this would be either to raise and catch a KeyError or return None if not found, instead of first checking whether it's in a dict then extracting.

Reply:
I am not that sure about that?
We don't want to return `None`, because the contract is that we only extract a variable if we expect it to be there no?
I find catching `KeyError` a bit yuck in my opinion. To be honest this should rather be an assert, as we don't expect this error to be visible to the user, as it should never be triggered whatever the user inputs are.